backends/core/models.py

```diff
@@ -267,105 +355,222 @@ def get_ready(self):
+      curr_status = cache.get_memcache_client().get(key, for_cas=True)
+      if curr_status != Job.STATUS.WAITING:
+        return None
+      elif cache.get_memcache_client().cas(key, Job.STATUS.RUNNING,
+          time=cache.MEMCACHE_DEFAULT_EXPIRATION_TIME_SECONDS):
+        # Update the database status.
+        self.update(status=Job.STATUS.RUNNING, status_changed_at=datetime.now())
+        return self.run()
+      else:
+        retries += 1
+
+    # Failed to start.
+    from core import cloud_logging
+    cloud_logging.logger.log_struct({
+        'labels': {
+            'pipeline_id': self.pipeline_id,
+            'job_id': self.id,
+            'worker_class': self.worker_class,
+        },
+        'log_level': 'ERROR',
+        'message': 'Cannot start the job after multiple retries.',
+    })
+    return None
+
   def run(self):
+    # TODO remove the enqueued_workers_count field (since we use memcache for that purpuse)
     self.enqueued_workers_count = 0
-    self.succeeded_workers_count = 0
-    self.failed_workers_count = 0
-    self.status = 'running'
-    self.status_changed_at = datetime.now()
     worker_params = dict([(p.name, p.val) for p in self.params])
-    self.enqueue(self.worker_class, worker_params)
+    return self.enqueue(self.worker_class, worker_params)
 
   def stop(self):
-    if self.status == 'waiting':
-      self.update(status='failed', status_changed_at=datetime.now())
+    self._cancel_tasks()
+    if self.get_status() == Job.STATUS.WAITING:
+      self.set_status(Job.STATUS.IDLE)
       return True
-    elif self.status == 'running':
-      self.update(status='stopping', status_changed_at=datetime.now())
+    elif self.get_status() == Job.STATUS.RUNNING:
+      self.set_status(Job.STATUS.STOPPING)
       return True
     return False
 
   def enqueue(self, worker_class, worker_params, delay=0):
-    if self.status != 'running':
-      return False
+    if self.get_status() != Job.STATUS.RUNNING:
+      return None
+
+    # Add a new task to the queue.
+    task_name = '%s_%s' % (self.pipeline.id, self.id)
+    escaped_task_name = re.sub(r'[^-_0-9a-zA-Z]', '-', task_name)
+    unique_task_name = '%s_%s' % (escaped_task_name, str(uuid.uuid4()))
     task_params = {
         'job_id': self.id,
         'worker_class': worker_class,
         'worker_params': json.dumps(worker_params),
+        'task_name': unique_task_name
     }
-    task_name = '%s_%s_%s' % (self.pipeline.name, self.name, self.worker_class)
-    escaped_task_name = re.sub(r'[^-_0-9a-zA-Z]', '-', task_name)
-    unique_task_name = '%s_%s' % (escaped_task_name, str(uuid.uuid4()))
     task = taskqueue.add(
         target='job-service',
         name=unique_task_name,
         url='/task',
         params=task_params,
         countdown=delay)
+
+    # Keep track of the running task name.
+    self._add_task_name_cache(unique_task_name)
+
+    # TODO remove these two lines when we will remove the enqueued_workers_count field
```

Comment:
Why do we keep enqueued_workers_count at all?

Reply:
Removed.